File: python/kindergarten-garden/kindergarten_garden.py
default_students = ['Alice', 'Bob', 'Charlie', 'David', 'Eve', 'Fred', 'Ginny', 'Harriet', 'Ileana', 'Joseph', 'Kincaid', 'Larry']
plants = { x[0]: x for x in ['Grass', 'Clover', 'Radishes', 'Violets'] }

class Garden(object):
    def __init__(self, garden, students=default_students):
    	students.sort()
    	g = garden.split()
        self._garden = { students[x/2]:self._get_plants(g, x) for x in range(0, len(g[0]), 2) }

    def _get_plants(self, g, x):
    	return [plants[g[0][x]], plants[g[0][x+1]], plants[g[1][x]], plants[g[1][x+1]]]

    def plants(self, student):
    	return self._garden[student]

File: javascript/kindergarten-garden/kindergarten-garden.js
const defaultStudents = ['Alice', 'Bob', 'Charlie', 'David', 'Eve', 'Fred', 'Ginny', 'Harriet', 'Ileana', 'Joseph', 'Kincaid', 'Larry'];
const plants = ['Grass', 'Clover', 'Radishes', 'Violets'].reduce((acc, p) => {
    acc[p[0]] = p.toLowerCase();
    return acc;
}, {});

class Garden {

    constructor(garden, students = defaultStudents) {
        let rows = garden
            .split('\n')
            .map((row) => {
                return [...row].map(c => plants[c]);
            });
        students.sort();
        
        rows[0]
            .reduce((acc, p, idx) => {
                let additions = [rows[0][idx], rows[0][idx + 1], rows[1][idx], rows[1][idx + 1]];
                if (idx % 2 === 0) {
                    acc.push(additions);
                }
                return acc;
            }, [])
            .forEach((section, idx) => {
                this[students[idx].toLowerCase()] = section;
            });
    }

}

module.exports = Garden;